src/core/ipfs/files.js

```diff
@@ -1,22 +1,57 @@
 'use strict'
 
-const Importer = require('ipfs-unixfs-engine').importer
-const Exporter = require('ipfs-unixfs-engine').exporter
+const unixfsEngine = require('ipfs-unixfs-engine')
+const Importer = unixfsEngine.Importer
+const Exporter = unixfsEngine.Exporter
 const UnixFS = require('ipfs-unixfs')
-const bs58 = require('bs58')
 const through = require('through2')
 const isStream = require('isstream')
 const promisify = require('promisify-es6')
+const Duplex = require('stream').Duplex
+const multihashes = require('multihashes')
 
 module.exports = function files (self) {
   return {
-    createAddStream: promisify((callback) => {
-      // TODO: wip
-      if (data === undefined) {
-        return new Importer(self._dagS)
+    createAddStream: (callback) => {
+      const i = new Importer(self._dagS)
+      const ds = new Duplex({ objectMode: true })
+
+      ds._read = (n) => {}
+      ds._write = (file, enc, next) => {
+        i.write(file)
+        next()
       }
-    }),
 
+      ds.end = () => {
+        i.end()
+      }
+
+      let counter = 0
+
+      i.on('data', (file) => {
+        counter++
+        self.object.get(file.multihash, (err, node) => {
+          if (err) {
+            return ds.emit('error', err)
+          }
+          ds.push({path: file.path, node: node})
+          counter--
+        })
+      })
+
+      i.on('end', () => {
+        function canFinish () {
+          if (counter === 0) {
+            ds.push(null)
+          } else {
+            setTimeout(canFinish, 100)
```

Comment:
Cause getting the object is an Async operation and the underlying stream from the importer won't care to wait that we have pushed all the objects to the upper stream and emit them all and then the 'end' event. This avoid a racing condition.